exploration/text-vs-code.md

```diff
@@ -0,0 +1,237 @@
+
+---
+
+Users editing a simple message and who wish to add an `input` or `local` annotiation
+to the message do not wish to reformat the message extensively.
+
+Users who have messages that include leading or trailing whitespace
+want to ensure that this whitespace is included in the translatable
+text portion of the message. Which whitespace characters are displayed at runtime
+should not be surprising.
+
+## Requirements
+
+Common things should be easy, uncommon things should be possible.
+
+Developers and translators should be able to read and write the syntax easily in a text editor.
+
+Translators (and their tools) are not software engineers, so we want our syntax
+to be as simple, robust, and non-fussy as possible.
+Multiple levels of complex nesting should be avoided,
+along with any constructs that require an excessive
+level of precision on the part of non-technical users.
+
+As MessageFormat 2 will be at best a secondary language to all its users,
+it should conform to user expectations and require as little learning as possible.
+
+The syntax should avoid footguns,
+in particular as it's passed through various tools during formatting.
+
+ASCII-compatible syntax. While support for non-ASCII characters for variable names,
+values, literals, options, and the like are important, the syntax itself should
+be restricted to ASCII characters. This allows the message to be parsed
+visually by humans even when embedded in a syntax that requires escaping.
+
+Whitespace is forgiving. We _require_ the minimum amount of whitespace and allow
+users to format or change unimportant whitespace as much as they want.
+This avoids the need for translators or tools to be super pedantic about
+formatting.
+
+## Constraints
+
+Limiting the range of characters that need to be escaped in plain text is important.
+
+The current syntax includes some plain-ascii keywords:
+`input`, `local`, `match`, and `when`.
+
+The current syntax and active proposals include some sigil + name combinations,
+such as `:number`, `$var`, `|literal|`, `+bold`, and `@attr`.
+
+The current syntax supports unquoted literal values as operands.
+
+Messages themselves are "simple strings" and must be considered to be a single
+line of text. In many containing formats, newlines will be represented as the local
+equivalent of `\n`.
+
+## Proposed Design
+
+### Start in text, encapsulate code, trim around statements
+
+Allow for message patterns to not be quoted.
+
+Encapsulate with `{…}` or otherwise distinguishing statements from
+the primarily unquoted translatable message contents.
+
+For messages with multiple variants,
+separate the variants using `when` statements.
+
+Trim whitespace between and around statements such as `input` and `when`,
+but do not otherwise trim any leading or trailing whitespace from a message.
+This allows for whitespace such as spaces and newlines to be used outside patterns
+to make a message more readable.
```

Comment:
I'm concerned about this resulting in surprising misbehavior... consider a developer making the following change, in which removing all inputs from a message that still starts and ends with a line feed would not be expected to affect whitespace:
`````diff
logOutMessage = ```
-{%input username}
-
-Log out {$username}?
+Log out?
```;
`````

Reply:
Ah, fair point. The issue here is that the above edit would add a new line to the message's start, yes? One way to avoid this would be to require a message with statements to not have leading whitespace. That might be a reasonable restriction.

Reply:
I agree with @gibson042's illustration being informative. I would have guessed (if I didn't know anything about MF2) that the "before" of `logOutMessage` included a newline. With some of the options here, the newline after `{%input $username}` is consumed, but not the blank link after that. In other options both newlines are consumed unless the user specifically quoted the pattern or whitespace.

Also, note that many formats would encode the example as:

```
// using our current syntax but without quoting the pattern
var logOutMessage = "{{{#input $username}\n\nLog out {$username}?}}
```